src/java.desktop/share/classes/sun/awt/image/SurfaceManager.java

```diff
@@ -89,47 +92,14 @@ public static void setManager(Image img, SurfaceManager mgr) {
         imgaccessor.setSurfaceManager(img, mgr);
     }
 
-    private volatile ConcurrentHashMap<Object,Object> cacheMap;
-
-    /**
-     * Return an arbitrary cached object for an arbitrary cache key.
-     * Other objects can use this mechanism to store cached data about
-     * the source image that will let them save time when using or
-     * manipulating the image in the future.
-     * <p>
-     * Note that the cache is maintained as a simple Map with no
-     * attempts to keep it up to date or invalidate it so any data
-     * stored here must either not be dependent on the state of the
-     * image or it must be individually tracked to see if it is
-     * outdated or obsolete.
-     * <p>
-     * The SurfaceData object of the primary (destination) surface
-     * has a StateTracker mechanism which can help track the validity
-     * and "currentness" of any data stored here.
-     * For convenience and expediency an object stored as cached
-     * data may implement the FlushableCacheData interface specified
-     * below so that it may be notified immediately if the flush()
-     * method is ever called.
-     */
-    public Object getCacheData(Object key) {
-        return (cacheMap == null) ? null : cacheMap.get(key);
-    }
-
     /**
-     * Store an arbitrary cached object for an arbitrary cache key.
-     * See the getCacheData() method for notes on tracking the
-     * validity of data stored using this mechanism.
+     * This map holds references to SurfaceDataProxy per given ProxyCache.
+     * Unlike ProxyCache, which contains SurfaceDataProxy objects per given SurfaceManager,
+     * this map does not prevent contained proxies from being garbage collected.
+     * Therefore, ProxyCache can be considered an "owning" container for the SurfaceDataProxy objects,
+     * and this map is just a weak mapping for the bookkeeping purposes.
      */
-    public void setCacheData(Object key, Object value) {
-        if (cacheMap == null) {
-            synchronized (this) {
-                if (cacheMap == null) {
-                    cacheMap = new ConcurrentHashMap<>(2);
-                }
-            }
-        }
-        cacheMap.put(key, value);
-    }
+    private final Map<ProxyCache, WeakReference<SurfaceDataProxy>> weakCache = new WeakHashMap<>(2);
```

Comment:
Ping ! on this .

Reply:
Because we only use it inside `synchronized` blocks. We need to do this explicitly because in one place it is needed for acquiring two monitors in particular order, and in second place it covers an iteration loop. Therefore there is no need to wrap it into synchronized wrapper.

Reply:
ok